### ISSUE TITLE
2.6.2 TRACKER_SCRIPT_NAME is still not working

### ISSUE THREAD

**Issue by MFYDev** [umami-software/umami#2264]
### Describe the Bug

Using the latest docker installation and when I input the customized name url, the log shows this

Failed to proxy https://192.168.144.3:3000/script.js Error: write EPROTO 58FB0D50FA7F0000:error:0A00010B:SSL routines:ssl3_get_record:wrong version number:../deps/openssl/openssl/ssl/record/ssl3_record.c:354:
    at WriteWrap.onWriteComplete [as oncomplete] (node:internal/stream_base_commons:94:16) {
  errno: -71,
  code: 'EPROTO',
  syscall: 'write'
}
Error: write EPROTO 58FB0D50FA7F0000:error:0A00010B:SSL routines:ssl3_get_record:wrong version number:../deps/openssl/openssl/ssl/record/ssl3_record.c:354:
    at WriteWrap.onWriteComplete [as oncomplete] (node:internal/stream_base_commons:94:16) {
  errno: -71,
  code: 'EPROTO',
  syscall: 'write'
}

### Database

PostgreSQL

**Comment by MFYDev** [comment]
The https:// is wrong, it should be http://. The proxy itself should be accessed by HTTP and not HTTPS even though the target URL is HTTPS. The proxy will nevertheless properly handle HTTPS connection and keep the end-to-end encryption. See [HTTP CONNECT method](https://en.wikipedia.org/wiki/HTTP_tunnel#HTTP_CONNECT_method) for details how this is done.

**Comment by MFYDev** [comment]
@mikecao

**Comment by mikecao** [comment]
This error looks like it's coming from your proxy server, not the app.

**Comment by MFYDev** [comment]
Yeah it is, it is behind Nginx and for the reverse proxy the config is like below:
location ^~ /
{
    proxy_pass http://localhost:3000;
    proxy_set_header Host myowndomain;
    proxy_set_header X-Real-IP $remote_addr;
    proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
    proxy_set_header REMOTE-HOST $remote_addr;
    proxy_set_header Upgrade $http_upgrade;
    proxy_set_header X-Forwarded-Proto $scheme;
    proxy_set_header Connection $connection_upgrade;
    proxy_http_version 1.1;
}

The whole domain is using SSL by the nginx, seems like it is correct?

**Comment by MFYDev** [comment]
I believe most of the websites will use https:// and docker will mostly run behind a web server. 

considering umami doesn't have a place to set ssl, seems like web server is the only place to set it?

so there might be somewhere within umami that can be enhanced for this case?

I am not fully familiar with umami's code so I am just assuming.

**Comment by MFYDev** [comment]
Btw the log which I pasted is from umami docker container

**Comment by saschwarz** [comment]
I'm having the same problem.
If I specify TRACKER_SCRIPT_NAME=umami.js
it worked in 2.6.0 but now in 2.6.2 I get the same error in the umami container as shown above.

**Comment by chrgeor** [comment]
I have the exact same error using traefik

**Comment by MFYDev** [comment]
Thank you guys for the confirmation on this.

Actually I started to use umami since a very long time ago. But honestly speaking self hosted Umami has given me two main problems most of the time that prevent me on using it.

1. When doing source installation and using MySQL 8.0 in local env, database can never be connected, check-db.js must has sth wrong
2. Tracker script name for docker after 2.6.0 has never worked properly.

However, considering you guys have this problem I do have a solution for you, this solution will not only works for Umami but also work for any other platform that you will use.

Add one more layer of reverse proxy in your external web server. For instance, me, I am using Nginx. Setting TRACKER_SCRIPT_NAME in Umami as `custom`. So I will add a reverse proxy for my umami like this:

```nginx

location /custom/
{
    proxy_pass http://localhost:3000/script.js;
    proxy_set_header Host t.mfy.world;
    proxy_set_header X-Real-IP $remote_addr;
    proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
    proxy_set_header REMOTE-HOST $remote_addr;
    proxy_set_header Upgrade $http_upgrade;
    proxy_set_header X-Forwarded-Proto $scheme;
    proxy_set_header Connection $connection_upgrade;
    proxy_http_version 1.1;
}
```

This will work on any tracking platform you use.

For me this time. I have experienced enough debug in Umami in my production use, and the problem happened really made my experience with Umami struggle. So I decided to switch to some other tracking platforms.

Good luck with you guys.

**Comment by mikecao** [comment]
In case it helps, here is my Nginx config:

```
upstream app {
  server umami:3000;
}

server {
  listen 80;
  listen [::]:80;

  server_name stats.mikecao.com;

  location / {
    proxy_set_header    Host $host;
    proxy_set_header    X-Real-IP $remote_addr;
    proxy_set_header    X-Forwarded-For $proxy_add_x_forwarded_for;
    proxy_set_header    X-Forwarded-Proto $scheme;
    proxy_pass          http://app;
    proxy_read_timeout  90;
    proxy_intercept_errors on;
  }
}
```

I am running the latest Umami v2.6.2 image and everything, include `TRACKER_SCRIPT_NAME` and `COLLECT_API_ENDPOINT` is working fine.

I am running Umami and Nginx under a single docker compose file.

**Comment by jekuer** [comment]
Setting the force_ssl env variable to 0 solved it somehow for me.

**Comment by MFYDev** [comment]
> force_ssl

This might be a good point, trying now

**Comment by MFYDev** [comment]
Just tried, force_ssl won't help.

> In case it helps, here is my Nginx config:
> 
> ```
> upstream app {
>   server umami:3000;
> }
> 
> server {
>   listen 80;
>   listen [::]:80;
> 
>   server_name stats.mikecao.com;
> 
>   location / {
>     proxy_set_header    Host $host;
>     proxy_set_header    X-Real-IP $remote_addr;
>     proxy_set_header    X-Forwarded-For $proxy_add_x_forwarded_for;
>     proxy_set_header    X-Forwarded-Proto $scheme;
>     proxy_pass          http://app;
>     proxy_read_timeout  90;
>     proxy_intercept_errors on;
>   }
> }
> ```
> 
> I am running the latest Umami v2.6.2 image and everything, include `TRACKER_SCRIPT_NAME` and `COLLECT_API_ENDPOINT` is working fine.
> 
> I am running Umami and Nginx under a single docker compose file.

Also tried what @mikecao said. still the same error, I have no idea why this happenes

**Comment by germanyague** [comment]
Same here. Upgrading to 2.6.2 gives a 500 Server error when trying to access the custom endpoint. The log says:

Failed to proxy https://172.19.0.3:3000/script.js Error: write EPROTO 585B98A7A57F0000:error:0A00010B:SSL routines:ssl3_get_record:wrong version number:../deps/openssl/openssl/ssl/record/ssl3_record.c:354:
    at WriteWrap.onWriteComplete [as oncomplete] (node:internal/stream_base_commons:94:16) {
  errno: -71,
  code: 'EPROTO',
  syscall: 'write'
}
Error: write EPROTO 585B98A7A57F0000:error:0A00010B:SSL routines:ssl3_get_record:wrong version number:../deps/openssl/openssl/ssl/record/ssl3_record.c:354:
    at WriteWrap.onWriteComplete [as oncomplete] (node:internal/stream_base_commons:94:16) {
  errno: -71,
  code: 'EPROTO',
  syscall: 'write'
}

**Comment by MFYDev** [comment]
I did not get enough time to look but now I just went through @mikecao 's config. He is not using https and that's the reason why it works for him...

**Comment by iuri-gg** [comment]
> Setting the force_ssl env variable to 0 solved it somehow for me.

this worked for me. Other pages besides custom script work fine with it. I don't think it is an nginx (proxy) issue. Something weird with how SSL is handled for custom script name.

**Comment by aaro-n** [comment]
The `TRACKER_SCRIPT_NAME` variable can work normally in the first few versions after umami 2.0, probably in March or April this year, because I need to download custom.js separately after updating the umami version each time (the `TRACKER_SCRIPT_NAME` variable is `fqik6NiqznM4a4h4bcw` ), the tracking code at that time was `<script async src="https://tg.ccc.cn/fqik6NiqznM4a4h4bcw.js" data-website-id="zzzzzz"></script>` instead of the current tracking code` <script async src="https://tg.ccc.cn/fqik6NiqznM4a4h4bcw" data-website-id="zzzzzz"></script>`, please note that the `TRACKER_SCRIPT_NAME` variable has not changed, I need to download umami’s js The file was uploaded to cloud storage and referenced in the website. After the `fqik6NiqznM4a4h4bcw.js` file could not be downloaded, I visited `https://tg.ccc.cn/custom.js` and saved the JS file as `fqik6NiqznM4a4h4bcw .js`, save it on the desktop and then upload it to cloud storage.
It has been a long time and the information provided is for reference only.

**Comment by rhbkweb** [comment]
Anyone having this issue using umami, installed using Cloudron.io control panel: https://www.cloudron.io/store/is.umami.cloudronapp.html ?

**Comment by mikecao** [comment]
@MFYDev You're right I wasn't using SSL. I went back and added SSL and indeed I can replicate the error.

To give some background, the difference between the normal app build and the docker build is the inclusion of a `docker/middleware.js` file is included in the docker image build. This is what handles the environment variable redirects. For some reason, it's causing this issue when SSL is involved. I am still trying to troubleshoot now.

**Comment by mikecao** [comment]
Ok, this seems to work for me, add an environment variable `HOSTNAME=0.0.0.0` and then start your Docker.

**Comment by saschwarz** [comment]
@mikecao Thanks for looking into this. That change worked for my environments.

**Comment by aaro-n** [comment]
Yes, set `HOSTNAME=0.0.0.0`, it works fine.

**Comment by jekuer** [comment]
confirmed

**Comment by mikecao** [comment]
Ok, if it's working for everyone, we will set the hostname automatically in the next image build.

**Comment by germanyague** [comment]
Working for me as well, yes.

**Comment by simonwiles** [comment]
+1 -- worked without on `v2.5.0` (multiple systems, each behind nginx), `v2.6.x` requires `HOSTNAME=0.0.0.0`.  Thanks!

**Comment by thomascenni** [comment]
Works for me also setting HOSTNAME=0.0.0.0.

**Comment by khromov** [comment]
Was seeing the same issue, `HOSTNAME=0.0.0.0` worked for me as well. Using Docker / CapRover.

**Comment by anseltaft** [comment]
The fix also worked for me behind a Traefik reverse proxy so thank you!

**Comment by chrgeor** [comment]
> I have the exact same error using traefik
> 
> edit:
> 
> The TRACKER_SCRIPT_NAME is used.
> 
> I logged in the container and wget the `http://172.24.0.2:3000/custom` and it seems to work
> 
> For some reason it (traefik or nginx) can't proxy the rewrite of the nextjs
> 
> edit2:
> 
> I temporarily fix it by adding a middleware to traefik
> 
> ```
>       - "traefik.http.middlewares.umami-replace-tracker.replacepathregex.regex=custom"
>       - "traefik.http.middlewares.umami-replace-tracker.replacepathregex.replacement=script.js"
> ```

I can confirm that this fix is not needed at v2.7.0 fd5d5c692e09fa5b756d0ca7b2a8c104e1c80b4a
